### PR TITLE
[REVIEW] Delete allocated host chunks

### PR DESCRIPTION
**Pull request by aucahuasi**
https://github.com/BlazingDB/blazingsql/issues/1317

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: engine/src/bmr/BufferProvider.cpp [hunk 1]
```diff
@@ -156,7 +158,22 @@ void allocation_pool::grow() {
...
+  const std::size_t idx = buffer->allocation->index;
   buffer->allocation->allocation_chunks.push(std::move(buffer));
+
+  // if when we add the chunk back to the allocation, it now has all the allocations it needs, they free the allocation
+  // BUT only free the allocation if its not the very first allocation
+  // freeing the allocation means freeing the memory and deleting it from the allocation pool
+  // make sure things are "locked" when doing that to make sure its thread safe
+  if (idx > 0) {
+    if (this->allocations.at(idx)->total_number_of_chunks == this->allocations.at(idx)->allocation_chunks.size()) {
+        auto it = this->allocations.begin();
```
Comment (williamBlazing) [inline engine/src/bmr/BufferProvider.cpp:170]: `this->allocations.begin()` is pointing at the first element. The element you would want to delete is `idx`.

File: engine/src/bmr/BufferProvider.cpp [hunk 2]
```diff
@@ -156,7 +158,22 @@ void allocation_pool::grow() {
 
 void allocation_pool::free_chunk(std::unique_ptr<blazing_allocation_chunk> buffer) {
   std::unique_lock<std::mutex> lock(in_use_mutex);
+  const std::size_t idx = buffer->allocation->index;
   buffer->allocation->allocation_chunks.push(std::move(buffer));
+
+  // if when we add the chunk back to the allocation, it now has all the allocations it needs, they free the allocation
+  // BUT only free the allocation if its not the very first allocation
+  // freeing the allocation means freeing the memory and deleting it from the allocation pool
+  // make sure things are "locked" when doing that to make sure its thread safe
```
Comment (williamBlazing) [inline engine/src/bmr/BufferProvider.cpp:167]: can you remove this line

File: engine/src/bmr/BufferProvider.cpp [hunk 3]
```diff
@@ -131,13 +131,15 @@ std::unique_ptr<blazing_allocation_chunk> allocation_pool::get_chunk() {
...
-
   // if this is the first growth (initializaton) then we want num_buffers, else we will just grow by half that.
   std::size_t num_new_buffers = this->buffer_counter == 0 ? this->num_buffers : this->num_buffers/2;
   allocations.push_back(std::make_unique<blazing_allocation>());
+  allocations.back()->index = this->allocations.size() - 1;
+  allocations.back()->total_number_of_chunks = num_new_buffers;
   auto last_index = allocations.size() -1;
   try{
     allocator->allocate((void **) &allocations[last_index]->data,num_new_buffers * buffer_size);
+    this->allocations[last_index]->total_number_of_chunks = num_new_buffers;
```
Comment (williamBlazing) [inline engine/src/bmr/BufferProvider.cpp:142]: You are setting `total_number_of_chunks` twice. Dont do that. Delete this one or the one above

**Review by williamBlazing — CHANGES_REQUESTED** [review]

File: engine/src/bmr/BufferProvider.cpp (hunk 1 above)
Comment (aucahuasi) [reply engine/src/bmr/BufferProvider.cpp:170]: the next line is an std::advance(it, idx) ... so this way I make the iterator point to that index

File: engine/src/bmr/BufferProvider.cpp (hunk 2 above)
Comment (aucahuasi) [reply engine/src/bmr/BufferProvider.cpp]: done

File: engine/src/bmr/BufferProvider.cpp (hunk 3 above)
Comment (aucahuasi) [reply engine/src/bmr/BufferProvider.cpp:140]: done

**Pull request merged**